clipper_admin_v2/clipper_admin/docker/docker_container_manager.py

```diff
@@ -0,0 +1,242 @@
+            # in same docker network as the query frontend
+            "CLIPPER_IP": self.query_frontend_name,
+            "CLIPPER_INPUT_TYPE": input_type,
+        }
+        self.extra_container_kwargs["labels"][
+            CLIPPER_MODEL_CONTAINER_LABEL] = "{name}:{version}".format(
+                name=name, version=version)
+        self.docker_client.containers.run(
+            repo, environment=env_vars, **self.extra_container_kwargs)
+
+    def set_num_replicas(self, name, version, input_type, repo, num_replicas):
+        current_replicas = self._get_replicas(name, version)
+        if len(current_replicas) < num_replicas:
+            num_missing = num_replicas - len(current_replicas)
+            logger.info(
+                "Found {cur} replicas for {name}:{version}. Adding {missing}".
+                format(
+                    cur=len(current_replicas),
+                    name=name,
+                    version=version,
+                    missing=(num_missing)))
+            for _ in range(num_missing):
+                self._add_replica(name, version, input_type, repo)
+        elif len(current_replicas) > num_replicas:
+            num_extra = len(current_replicas) - num_replicas
+            logger.info(
+                "Found {cur} replicas for {name}:{version}. Removing {extra}".
+                format(
+                    cur=len(current_replicas),
+                    name=name,
+                    version=version,
+                    extra=(num_extra)))
+            while len(current_replicas) > num_replicas:
+                cur_container = current_replicas.pop()
+                cur_container.stop()
+
+    def get_logs(self, logging_dir):
+        containers = self.docker_client.containers.list(
+            filters={"label": CLIPPER_DOCKER_LABEL})
+        logging_dir = os.path.abspath(os.path.expanduser(logging_dir))
+
+        log_files = []
+        if not os.path.exists(logging_dir):
+            os.makedirs(logging_dir)
+            logger.info("Created logging directory: %s" % logging_dir)
+        for c in containers:
+            log_file_name = "image_{image}:container_{id}.log".format(
+                image=c.image.short_id, id=c.short_id)
+            log_file = os.path.join(logging_dir, log_file_name)
+            with open(log_file, "w") as lf:
+                lf.write(c.logs(stdout=True, stderr=True))
+            log_files.append(log_file)
+        return log_files
+
+    def stop_models(self, model_name=None, keep_version=None):
+        containers = self.docker_client.containers.list(
+            filters={"label": CLIPPER_MODEL_CONTAINER_LABEL})
+        for c in containers:
+            c_name, c_version = c.labels[CLIPPER_MODEL_CONTAINER_LABEL].split(
+                ":")
+            if model_name is not None and model_name == c_name:
+                if keep_version is None or keep_version != c_version:
+                    c.stop()
```

Comment:
Well I'm going to rewrite this function anyway, but this misses the case where model_name is None and we want to remove all model containers